pkg/piecestore/rpc/client/readerwriter.go

```diff
@@ -94,22 +97,10 @@ func NewStreamReader(signer *Client, stream pb.PieceStoreRoutes_RetrieveClient,
 	}
 
 	go func() {
-		<-sr.c
-		for sr.totalRead < sr.max {
-			sr.allocSendCond.L.Lock()
-			fmt.Println("Locking ba send")
-			for sr.allocated-sr.totalRead >= int64(4*signer.bandwidthMsgSize) {
-				sr.allocSendCond.Wait() // Wait for the bandwidth loop to get a new bandwidth allocation
-			}
-			fmt.Println("Unlocking ba send")
-			sr.allocSendCond.L.Unlock()
-
-			fmt.Println(sr.allocated, sr.totalRead)
-
-			if sr.max-sr.allocated < int64(signer.bandwidthMsgSize) {
-				sr.allocated += sr.max - sr.allocated
-			} else {
-				sr.allocated += int64(signer.bandwidthMsgSize)
+		for {
+			err := sr.throttle.ProduceAndWaitUntilBelow(int64(signer.bandwidthMsgSize), max)
+			if err != nil {
+				// TODO what should we do about an error here?
```

Comment:
Just return here and use `sr.throttle.Err()` to return the failure eventually in `StreamReader`.